### PR TITLE
Default phone

### PR DESCRIPTION
**Pull request by yramanchuk**
Implement phone Auth flow Phone number configuration parameter.
Related issue https://github.com/firebase/FirebaseUI-Android/issues/887

### CODE REVIEW

**Comment by SUPERCILEX** [comment]
@yramanchuk I believe the result of that issue was that FirebaseUI isn't going to support this kind of flow. Could you explain why you already have a phone number? The user should be giving that to FirebaseUI in the auth flow, not to you directly. Hope this clears things up! 😄

**Comment by ashwinraghav** [comment]
@yramanchuk I think the issue that is linked to may not be completely comparable.

@SUPERCILEX The problem we are addressing is rooted in the fact that Firebase UI currently does not support 2FA (Say Facebook + Phone Auth). To work around this, apps that wish to use phone as the second factor can use Facebook login on their own outside Firebase UI and then use Firebase UI for the second factor phone auth.
This change helps apps that "know" the second factor phone number given a Facebook account, particularly when a user is logging into their secondary device like a tablet. Hope that helps.

**Comment by yramanchuk** [comment]
@ashwinraghav Thanks for the clarification.
@SUPERCILEX here's another example: application may have multi-dialog account creation flow and they have already requested user to enter his phone number, so they would like to show phone verification dialog with pre-filled phone number.

Here's related FirebaseUI-iOS issue: https://github.com/firebase/FirebaseUI-iOS/issues/306 
This request is required for the feature parity: https://github.com/firebase/FirebaseUI-iOS/pull/334

File: auth/src/main/java/com/firebase/ui/auth/ui/idp/AuthMethodPickerActivity.java [hunk 1]
```diff
@@ -114,7 +115,8 @@ private void populateIdpList(List<IdpConfig> providers) {
                     mProviders.add(new EmailProvider(this, getFlowParams()));
                     break;
                 case AuthUI.PHONE_VERIFICATION_PROVIDER:
-                    mProviders.add(new PhoneProvider(this, getFlowParams()));
+                    PhoneIdpConfig phoneIdpConfig = (PhoneIdpConfig) idpConfig;
```
Comment (ashwinraghav) [inline auth/src/main/java/com/firebase/ui/auth/ui/idp/AuthMethodPickerActivity.java]: Would this break older clients, breaking backward compatibility?

File: auth/src/main/java/com/firebase/ui/auth/AuthUI.java [hunk 2]
```diff
@@ -378,6 +368,79 @@ public IdpConfig build() {
         }
     }
 
+    public static class PhoneIdpConfig extends IdpConfig {
```
Comment (ashwinraghav) [inline auth/src/main/java/com/firebase/ui/auth/AuthUI.java]: I feel like introducing a config per for phone auth provider leaves me desiring symmetry for such a config in other providers. 

Also, the type safety that this implementation gives is hard to enforce in a backward compatible way. This leads to unsafe typecasting like the one made in https://github.com/firebase/FirebaseUI-Android/blob/d1da32786f33f7a5552d3a3c5c778eb58c24305a/auth/src/main/java/com/firebase/ui/auth/ui/idp/AuthMethodPickerActivity.java#L118

My suggestion would be to introduce a Bundle parameter into the existing IdpConfig that yields an API that would look like:

Bundle b = new Bundle();
b.putString(AuthUI.PHONE_VERIFICATION_PROVIDER.DEFAULT_PHONE_EXTRA, "+16478488383");

new AuthUI.IdpConfig.Builder(AuthUI.PHONE_PROVIDER).setParams(bundle).build()) and wire it into the providers from that point on.

Comment (yramanchuk) [reply auth/src/main/java/com/firebase/ui/auth/AuthUI.java]: @ashwinraghav Thhank's for the suggestion. I was thinking about extending existing IdpConfig but in that case were are adding changes to all configs (Google, Facebook) and they they don't need that parameter.

Anyway I think your solution is more elegant. I just want to propose to use List of Bundles instead of just Bundle:

AuthUI.IdpConfig.Builder(AuthUI.PHONE_PROVIDER).setParams(List<Bundle>).build())

In that case we can pass multiple params.
@ashwinraghav what do you think?

**Comment by samtstern** [comment]
@yramanchuk FYI I changed the target branch to the current development branch, `version-2.3.1-dev`.

**Comment by samtstern** [comment]
Note that the issue of IDP-specific configuration has been discussed here as well:
https://github.com/firebase/FirebaseUI-Android/issues/806

File: app/src/main/java/com/firebase/uidemo/auth/AuthUiActivity.java [hunk 3]
```diff
@@ -337,8 +337,14 @@ private int getSelectedLogo() {
         }
 
         if (mUsePhoneProvider.isChecked()) {
+            // Use the params Bundle to provide a default phone number that will populate the
+            // phone number field in PhoneVerificationActivity if the number is known in advance.
+            Bundle params = new Bundle();
+            params.putString(AuthUI.PHONE_VERIFICATION_DEFAULT_PHONE_EXTRA, "+13099126574");
```
Comment (samtstern) [inline app/src/main/java/com/firebase/uidemo/auth/AuthUiActivity.java]: This is a minor point, but I think putting a random phone number into the sample will cause confusion.  Many people will want to use this sample to test phone auth with their real number.

This type of example snippet belongs in the README.

File: auth/src/main/java/com/firebase/ui/auth/util/signincontainer/SignInDelegate.java [hunk 4]
```diff
@@ -248,8 +248,12 @@ private void startAuthMethodChoice() {
                     break;
                 case PhoneAuthProvider.PROVIDER_ID:
                     // Go directly to phone flow
+                    String phone = idpConfigs
+                            .get(0)
+                            .getParams()
+                            .getString(AuthUI.PHONE_VERIFICATION_DEFAULT_PHONE_EXTRA);
                     startActivityForResult(
-                            PhoneVerificationActivity.createIntent(getContext(), flowParams, null),
+                            PhoneVerificationActivity.createIntent(getContext(), flowParams, phone),
```
Comment (samtstern) [inline auth/src/main/java/com/firebase/ui/auth/util/signincontainer/SignInDelegate.java]: This `createIntent` call uses `ExtraConstants.EXTRA_PHONE`.  We should unify the values between this and AuthUI.PHONE_VERIFICATION_DEFAULT_PHONE_EXTRA.

I'd suggest:

```java

class AuthUI {
  // ...
  public static final String EXTRA_DEFAULT_PHONE = ExtraConstants.EXTRA_PHONE;
  // ...
}
```

File: auth/src/main/java/com/firebase/ui/auth/ui/idp/AuthMethodPickerActivity.java (hunk 1 above)
Comment (samtstern) [inline auth/src/main/java/com/firebase/ui/auth/ui/idp/AuthMethodPickerActivity.java]: This class `PhoneIdpConfig` is used here but I don't see it defined in this PR anymore.  I think changes to this file need to be revisited.

**Comment by samtstern** [comment]
@SUPERCILEX the main reason for supporting this flow is feature parity with what Digits used to offer.  We are telling all Digits developers to move to FirebaseUI and some of them noticed this missing feature.

You are right though, it was not a flow we originally intended to support.  Ashwin and Yury have provided some examples where it can be very useful though.

File: app/src/main/java/com/firebase/uidemo/auth/AuthUiActivity.java (hunk 3 above)
Comment (jstimes) [reply app/src/main/java/com/firebase/uidemo/auth/AuthUiActivity.java]: Moved to README, let me know if you don't think it's in an appropriate location.

File: auth/src/main/java/com/firebase/ui/auth/util/signincontainer/SignInDelegate.java (hunk 4 above)
Comment (jstimes) [reply auth/src/main/java/com/firebase/ui/auth/util/signincontainer/SignInDelegate.java]: That's reasonable, made your suggested change.

File: auth/src/main/java/com/firebase/ui/auth/ui/idp/AuthMethodPickerActivity.java (hunk 1 above)
Comment (jstimes) [reply auth/src/main/java/com/firebase/ui/auth/ui/idp/AuthMethodPickerActivity.java]: Oops, gone now.

File: app/src/main/java/com/firebase/uidemo/auth/AuthUiActivity.java [hunk 5]
```diff
@@ -338,7 +338,8 @@ private int getSelectedLogo() {
 
         if (mUsePhoneProvider.isChecked()) {
             selectedProviders.add(
-                    new IdpConfig.Builder(AuthUI.PHONE_VERIFICATION_PROVIDER).build());
+                    new IdpConfig.Builder(AuthUI.PHONE_VERIFICATION_PROVIDER)
+                            .build());
```
Comment (SUPERCILEX) [inline app/src/main/java/com/firebase/uidemo/auth/AuthUiActivity.java]: Might as well undo the changes in this file since they're leftovers from previous commits.

File: auth/src/main/java/com/firebase/ui/auth/AuthUI.java [hunk 6]
```diff
@@ -329,15 +347,27 @@ public int hashCode() {
...
         public String toString() {
+            StringBuilder paramsValueBuilder = new StringBuilder();
+            paramsValueBuilder.append("{");
+            for (String key : mParams.keySet()) {
+                paramsValueBuilder.append(key);
+                paramsValueBuilder.append("=");
+                paramsValueBuilder.append(mParams.get(key).toString());
+                paramsValueBuilder.append(",");
+            }
+            paramsValueBuilder.append("}");
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/AuthUI.java]: Doesn't `Bundle` already have a `toString()` with very similar behavior?

File: auth/src/main/java/com/firebase/ui/auth/provider/PhoneProvider.java [hunk 7]
```diff
@@ -34,8 +37,16 @@ public int getButtonLayout() {
 
     @Override
     public void startLogin(Activity activity) {
+
+        String phone = null;
+        for (AuthUI.IdpConfig idpConfig : mFlowParameters.providerInfo) {
+            if (idpConfig.getProviderId().equals(AuthUI.PHONE_VERIFICATION_PROVIDER)) {
+                phone = idpConfig.getParams().getString(EXTRA_DEFAULT_PHONE);
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/provider/PhoneProvider.java]: To be consistent, we shouldn't statically import this and instead use `AuthUI.EXTRA_DEFAULT_PHONE`.

File: auth/src/main/java/com/firebase/ui/auth/util/signincontainer/SignInDelegate.java [hunk 8]
```diff
@@ -248,8 +248,12 @@ private void startAuthMethodChoice() {
                     break;
                 case PhoneAuthProvider.PROVIDER_ID:
                     // Go directly to phone flow
+                    String phone = idpConfigs
+                            .get(0)
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/util/signincontainer/SignInDelegate.java]: This same logic is used above so we should probably extract a `firstIdpConfig` and replace `firstProvider` with the method call.

File: auth/src/main/java/com/firebase/ui/auth/AuthUI.java (hunk 6 above)
Comment (SUPERCILEX) [reply auth/src/main/java/com/firebase/ui/auth/AuthUI.java]: Yeah, just check the source and `Bundle#toString()` calls `mMap.toString()`. I'm assuming the formatting will be a little bit different, but the output should be effectively the same.

File: auth/src/main/java/com/firebase/ui/auth/AuthUI.java [hunk 9]
```diff
@@ -269,15 +276,21 @@ public SignInIntentBuilder createSignInIntentBuilder() {
...
+                Bundle params) {
             mProviderId = providerId;
             mScopes = Collections.unmodifiableList(scopes);
+            mParams = params;
         }
 
         private IdpConfig(Parcel in) {
             mProviderId = in.readString();
             mScopes = Collections.unmodifiableList(in.createStringArrayList());
+            mParams = in.readBundle();
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/AuthUI.java]: This should be `in.readBundle(getClass().getClassLoader())` to make the build pass. Since we're just passing in strings right now we don't actually need this, but it'll ensure we're future proof (and make lint happy).

File: app/src/main/java/com/firebase/uidemo/auth/AuthUiActivity.java (hunk 5 above)
Comment (jstimes) [reply app/src/main/java/com/firebase/uidemo/auth/AuthUiActivity.java]: Reverted.

File: auth/src/main/java/com/firebase/ui/auth/AuthUI.java (hunk 6 above)
Comment (jstimes) [reply auth/src/main/java/com/firebase/ui/auth/AuthUI.java]: Ah, did not realize that. Switched to the Bundle toString.

File: auth/src/main/java/com/firebase/ui/auth/provider/PhoneProvider.java (hunk 7 above)
Comment (jstimes) [reply auth/src/main/java/com/firebase/ui/auth/provider/PhoneProvider.java]: Fixed

File: auth/src/main/java/com/firebase/ui/auth/util/signincontainer/SignInDelegate.java (hunk 8 above)
Comment (jstimes) [reply auth/src/main/java/com/firebase/ui/auth/util/signincontainer/SignInDelegate.java]: Good point, done.

File: auth/src/main/java/com/firebase/ui/auth/AuthUI.java (hunk 9 above)
Comment (jstimes) [reply auth/src/main/java/com/firebase/ui/auth/AuthUI.java]: Ah, just noticed the CONTRIBUTING.md file. Apologies - just beginning contribution to the library and obviously wasn't paying full attention. Made the change and build should be passing now. Thanks for the review by the way!

File: auth/src/main/java/com/firebase/ui/auth/AuthUI.java [hunk 10]
```diff
@@ -332,12 +350,14 @@ public String toString() {
             return "IdpConfig{" +
                     "mProviderId='" + mProviderId + '\'' +
                     ", mScopes=" + mScopes +
+                    ", mParams=" + mParams.toString() +
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/AuthUI.java]: One last nit! 😄 We don't need the `toString()` call, `mParams` alone is just fine.

Comment (jstimes) [reply auth/src/main/java/com/firebase/ui/auth/AuthUI.java]: Removed.

**Comment by samtstern** [comment]
I will merge this once Travis is happy.
